### DISCUSSION TITLE
Cluster POIs on low zoom levels

### DISCUSSION THREAD

**Discussion by georgbachmann** [maplibre/martin#1020] [Q&A]
I am very new to martin, like the possibilities I think I might have with it :) 

So I have a map where I want to display POIs, but those should be clustered… MapLibre does not support clustering of a vector-tile-source it seems, right?!? I guess I have to do clustering then on the martin side of things, right?

Currently we have around 200.000 entries in our table, so not that huge… But doing it on the fly would still be a bit much CPU work I guess?!?
I guess I would need to do a per zoom level e.g. ST_ClusterDBScan and group by that using the function functionality from martin?

But as i am again very new to martin I would love to get feedback and ideas. Maybe there are many better ways of doing it that I simply overlooked.

**Comment by nyurik** [comment]
@georgbachmann I would recommend evaluating the performance of the ST_ClusterDBScan first - if it works for your usage, than you won't need to do anything else.  Plus you can setup varnish or nginx(?) or some other reverse proxy to cache responses, plus use CDN. This way it will be cached most of the time regardless.

If the solution is still too slow, you can use various database tricks to make it faster. For example, you can create a materialized view that pre-generates clustered tiles for low zooms (e.g. 0-9 is only 350,000 tiles that you can store in a single table). Then write a SQL function that either goes to that cache table (for low zooms), or uses your POIs directly for higher zooms.  This way you will only work with a tiny subset of the POIs in real time, and go to the materialized view for pre-generated tiles.

**Comment by KiwiKilian** [comment]
@georgbachmann did you proceed with this approach? If so, are you willing to share your SQL function for Martin?

**Reply by georgbachmann** [reply]
@KiwiKilian I played around with both ST_ClusterKMeans and ST_ClusterDBSCAN and they did cluster properly... but I did not like that approach after all. First up it was not as performant as I would have hoped. I guess with proper caching that should not be a huge problem. But what I did not like was the per-tile clustering, so it did not look as good as i was hoping for.
I am now using Martin to just serve a vector layer with most of my points (it's not thaaaaat many and we limit per tile)... and do the clustering on the client-side (using maplibre maps)
Hope that helps

**Reply by Fabioni** [reply]
@georgbachmann  can you explain how you do this on the client side? Maybe just sharing the maplibre code?

**Reply by georgbachmann** [reply]
@Fabioni i have a source that contains all my data. It's limited, so it does not explode :-) 
Then to display the it, i create a maplibre source like so:

```
        let source = MGLShapeSource(identifier: "<my-identifier-here>", features: annotations, options: [
            .clustered: annotations.count > 1, .clusterRadius: 45, .maximumZoomLevelForClustering: 13.75
        ])
```

and then i add some more style layers to display the data...
e.g. the one for the number:

```
        let numbersLayer = MGLSymbolStyleLayer(identifier: "<my-identifier-here-again>.numberLabel", source: source)
        numbersLayer.textColor = NSExpression(forConstantValue: UIColor.white)
        numbersLayer.textFontSize = NSExpression(forConstantValue: 12)
        numbersLayer.text = NSExpression(format: "CAST(point_count, 'NSString')")
        numbersLayer.predicate = NSPredicate(format: "cluster == YES")
```

but there are several layers to make everything look pretty... e.g. a background circle and so on... 

hope that helps you get started...
